server/aws-lsp-identity/src/sso/deviceCode/deviceCodeFlow.ts

```diff
@@ -0,0 +1,160 @@
+    AsyncTimeoutError,
+    getSsoOidc,
+    SsoFlowParams,
+    throwOnInvalidClientRegistration,
+    throwOnInvalidSsoSession,
+    UpdateSsoTokenFromCreateToken,
+} from '../utils'
+import { AwsError } from '@aws/lsp-core'
+import {
+    AuthorizationPendingException,
+    CreateTokenCommandOutput,
+    SlowDownException,
+    StartDeviceAuthorizationCommandOutput,
+} from '@aws-sdk/client-sso-oidc'
+
+const deviceGrantType = 'urn:ietf:params:oauth:grant-type:device_code'
+const defaultExpiration = 10 * 60 * 1000
+const responseTimeout = defaultExpiration
+const backoffDelayMs = 5000
+
+export async function deviceCodeFlow(params: SsoFlowParams): Promise<SSOToken> {
+    throwOnInvalidClientRegistration(params.clientRegistration)
+    throwOnInvalidSsoSession(params.ssoSession)
+
+    using oidc = getSsoOidc(params.ssoSession.settings.sso_region)
+
+    const authorization = await oidc.startDeviceAuthorization({
+        startUrl: params.ssoSession.settings.sso_start_url,
+        clientId: params.clientRegistration.clientId,
+        clientSecret: params.clientRegistration.clientSecret,
+    })
+    const expiresAt = new Date(
+        (authorization.expiresIn ? authorization.expiresIn * 1000 : defaultExpiration) + Date.now()
+    )
+
+    // Sanity check
+    if (!authorization.verificationUriComplete || !authorization.userCode || !authorization.deviceCode) {
+        throw new AwsError('Device code authorization is invalid.', AwsErrorCodes.E_INVALID_SSO_CLIENT)
+    }
+
+    // Show auth code and URL to the user
+    params.observability.logging.log(`Requesting ${params.clientName} to open SSO OIDC authorization login website.`)
+    await showOnClient(params.handlers, authorization)
+
+    // Wait for user to authenticate
+    params.observability.logging.log('Waiting for user to verify authorization code...')
+    let result
+    try {
+        result = await pollAuthorization(
+            () =>
+                oidc.createToken({
+                    clientId: params.clientRegistration.clientId,
+                    clientSecret: params.clientRegistration.clientSecret,
+                    deviceCode: authorization.deviceCode,
+                    grantType: deviceGrantType,
+                }),
+            expiresAt,
+            params.token,
+            authorization.interval ? authorization.interval * 1000 : undefined
+        )
+    } finally {
+        // Tell the client to close the progress bar regardless
+        params.handlers.showProgress(GetSsoTokenProgressType, GetSsoTokenProgressToken, {
+            state: GetSsoTokenProgressState.Complete,
+        })
+    }
+
+    params.observability.logging.log('Storing and returning created SSO token.')
+    return UpdateSsoTokenFromCreateToken(result, params.clientRegistration, params.ssoSession)
+}
+
+function pollAuthorization(
+    fn: () => Promise<CreateTokenCommandOutput>,
+    expiresAt: Date,
+    token: CancellationToken,
+    intervalMs: number = backoffDelayMs
```

Comment:
nit: For consistency with other identity code, consider using "Millis" instead of "Ms".